### PR TITLE
Group multiline expressions with parens, fix multiline string escaping

### PR DESCRIPTION
**Pull request by ryanashcraft**
My attempt at a follow-up from https://github.com/prettier/plugin-python/pull/72 – surrounding the printed output for BinOp/BoolOp nodes with parens instead of using escaped lines.

This also fixes multiline strings not being escaped properly when on the right-hand side of an assignment or return expression.

### CODE REVIEW

File: tests/python_comments/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -15,7 +15,7 @@ if a:  # a
         # a
     )
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-if a: # a
+if (a): # a
```
Comment (ryanashcraft) [inline tests/python_comments/__snapshots__/jsfmt.spec.js.snap]: Seems the comment here is triggering the `ifBreak` node to be printed, even though it's on the same line. Not sure what I'm doing wrong here or if this is just another bug.

File: src/printer/index.js [hunk 2]
```diff
@@ -565,6 +571,18 @@ function genericPrint(path, options, print) {
     }
 
     case "Expr": {
+      if (n.value) {
+        if (n.value.ast_type === "BoolOp" || n.value.ast_type === "BinOp") {
+          return group(
+            concat([
```
Comment (FuegoFro) [inline src/printer/index.js]: nit: if you wanted could use `groupConcat` here (and elsewhere)

File: src/printer/index.js [hunk 3]
```diff
@@ -565,6 +571,18 @@ function genericPrint(path, options, print) {
...
+              ifBreak("("),
+              indent(concat([softline, path.call(print, "value")])),
```
Comment (FuegoFro) [inline src/printer/index.js]: nit: if you wanted could use `indentConcat` here (and elsewhere)

File: src/printer/index.js [hunk 4]
```diff
@@ -565,6 +571,18 @@ function genericPrint(path, options, print) {
...
+              softline,
+              ifBreak(")")
+            ])
+          );
+        }
+      }
```
Comment (FuegoFro) [inline src/printer/index.js]: This looks like a pretty common pattern (check if it's a `BoolOp` or `BinOp` then do this group and indent with the parens). Would it make sense to pull it out to a function?

File: tests/python_comments/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (FuegoFro) [reply tests/python_comments/__snapshots__/jsfmt.spec.js.snap]: Hm interesting. Yeah I think when I looked into this the comment has a `hardline` in it, which triggers the `ifBreak` for the line. I'd be okay with leaving this for now and coming back to address it in a later change.

**Comment by ryanashcraft** [comment]
@FuegoFro Thanks for reviewing. I fixed up all `indentConcat`s and `groupConcat`s.

Regarding extracting the parens logic into reusable functions: I was a hesitant to do that initially because I didn't want to build an abstraction that might break down in the future. I also observed that the [JS ESTree printer](https://github.com/prettier/prettier/blob/master/src/language-js/printer-estree.js) does not reuse the same printing logic for similar situations (e.g. "ReturnStatement" and "AssignmentExpression" nodes). However, I made an attempt with my [latest commit](https://github.com/prettier/plugin-python/pull/79/commits/df1ab8704053b3da225481fb59f421779ad28208) so we can evaluate the tradeoffs. What do you think?

**Comment by ryanashcraft** [comment]
@FuegoFro I'm alright with keeping it DRY for now. Shouldn't be hard to inline later, if needed. Thanks for reviewing!

**Pull request merged**